pkg/machine/cleanup.go

```diff
@@ -0,0 +1,66 @@
+package machine
+
+import (
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+
+	"github.com/sirupsen/logrus"
+)
+
+type CleanupCallback struct {
+	Funcs []func() error
+	mu    sync.Mutex
+}
+
+func (c *CleanupCallback) Clean(err *error) {
+	// do not remove created files if the init is successful
+	if *err == nil {
+		return
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	for _, cleanfunc := range c.Funcs {
+		if err := cleanfunc(); err != nil {
+			logrus.Error(err)
+		}
+	}
+}
+
+func (c *CleanupCallback) CleanOnSignal(err *error) {
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, os.Interrupt, syscall.SIGTERM)
+
+	_, ok := <-ch
+	if !ok {
+		return
+	}
+	// if err exists, then we let the deferred clean function handle the cleanup
+	// this prevents a race where both defer and cleanonsignal tries to handle cleanups
+	if *err != nil {
+		return
+	}
+
+	c.mu.Lock()
+	for _, cleanfunc := range c.Funcs {
+		if err := cleanfunc(); err != nil {
+			logrus.Error(err)
+		}
+	}
+	c.mu.Unlock()
```

Comment:
You can achieve exclusive ownership under lock by resetting the value to nil (or alternatively an empty array) like so:

```suggestion
	c.mu.Lock()
	// Claim exclusive usage by copy and resetting to nil
	funcs := c.Funcs
	c.Funcs = nil
	c.mu.Unlock()
	
	// Cleanup functions can now exclusively be run
	for _, cleanfunc := range funcs {
		if err := cleanfunc(); err != nil {
			logrus.Error(err)
		}
	}
```